File: ios/Yard/Yard/Core/Models/Listing.swift
import Foundation

enum ListingCondition: String, Codable, CaseIterable, Sendable {
    case new
    case likeNew = "like_new"
    case good
    case fair

    var displayName: String {
        switch self {
        case .new: "New"
        case .likeNew: "Like new"
        case .good: "Good"
        case .fair: "Fair"
        }
    }
}

enum ListingStatus: String, Codable, Sendable {
    case draft
    case pendingModeration = "pending_moderation"
    case active
    case reserved
    case sold
    case archived
    case rejected
    case removed

    var displayName: String {
        switch self {
        case .draft: "Draft"
        case .pendingModeration: "In review"
        case .active: "Active"
        case .reserved: "Reserved"
        case .sold: "Sold"
        case .archived: "Archived"
        case .rejected: "Needs changes"
        case .removed: "Removed"
        }
    }
}

struct Listing: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    let sellerID: UUID
    let title: String
    let description: String
    let categoryID: UUID
    let subcategoryID: UUID?
    let categoryName: String
    let subcategoryName: String?
    let priceCents: Int
    let isFree: Bool
    let condition: ListingCondition
    let status: ListingStatus
    let pickupZone: String
    let imageURL: URL?
    let publishedAt: Date?
    let viewCount: Int
    let saveCount: Int

    var formattedPrice: String {
        guard !isFree else { return "Free" }
        return (Double(priceCents) / 100).formatted(.currency(code: "USD").precision(.fractionLength(0)))
    }
}

struct ListingPage: Codable, Sendable {
    let items: [Listing]
    let total: Int
    let limit: Int
    let offset: Int
}

struct ListingFilters: Equatable, Sendable {
    var query = ""
    var category: String?
    var subcategory: String?
    var condition: ListingCondition?
    var minimumPriceCents: Int?
    var maximumPriceCents: Int?
    var freeOnly = false
    var pickupZone: String?
    var maximumAgeDays: Int?
    var sort = ListingSort.recommended
}

enum ListingSort: String, CaseIterable, Sendable {
    case recommended
    case newest
    case priceAscending = "price_asc"
    case priceDescending = "price_desc"
    case closest

    var displayName: String {
        switch self {
        case .recommended: "Recommended"
        case .newest: "Newest"
        case .priceAscending: "Price: low to high"
        case .priceDescending: "Price: high to low"
        case .closest: "Closest pickup match"
        }
    }
}

File: backend/app/schemas/listing.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.listing import ListingCondition, ListingStatus
from app.schemas.listing_image import ListingImageRead


class ListingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    seller_id: uuid.UUID
    title: str
    description: str
    category_id: uuid.UUID
    subcategory_id: uuid.UUID | None
    category_name: str
    subcategory_name: str | None
    price_cents: int
    is_free: bool
    condition: ListingCondition
    status: ListingStatus
    pickup_zone: str
    image_url: str | None
    images: list[ListingImageRead]
    published_at: datetime | None
    view_count: int
    save_count: int


class ListingPage(BaseModel):
    items: list[ListingRead]
    total: int
    limit: int
    offset: int


class ListingQuery(BaseModel):
    query: str | None = Field(default=None, max_length=120)
    category: str | None = None
    subcategory: str | None = None
    condition: ListingCondition | None = None
    min_price_cents: int | None = Field(default=None, ge=0)
    max_price_cents: int | None = Field(default=None, ge=0)
    free_only: bool = False
    pickup_zone: str | None = None
    max_age_days: int | None = Field(default=None, ge=1, le=365)
    sort: str = "recommended"
    limit: int = Field(default=30, ge=1, le=100)
    offset: int = Field(default=0, ge=0)


class ListingDraftCreate(BaseModel):
    title: str = Field(min_length=3, max_length=140)
    description: str = Field(min_length=3, max_length=5_000)
    category_id: uuid.UUID
    subcategory_id: uuid.UUID | None = None
    price_cents: int = Field(ge=0, le=10_000_000)
    is_free: bool = False
    condition: ListingCondition
    pickup_zone: str = Field(min_length=2, max_length=100)


class ListingSafeUpdate(BaseModel):
    price_cents: int | None = Field(default=None, ge=0, le=10_000_000)
    is_free: bool | None = None
    condition: ListingCondition | None = None
    pickup_zone: str | None = Field(default=None, min_length=2, max_length=100)

    @model_validator(mode="after")
    def require_change(self) -> "ListingSafeUpdate":
        if not self.model_fields_set:
            raise ValueError("Provide at least one field to update")
        return self
